### ISSUE TITLE
Login to MySQL [Docker] fails using libmysql - Login works only after accessing MySQL with other clients first

### ISSUE THREAD

**Issue by Maxiride** [HeidiSQL/HeidiSQL#980]
**TL;DR**
Observed behavior is that HeidiSQL is able to login to the MySQL instance only after a successful login with another client (MySQL Workbench and dbForge tested), subsequent logins succeeds.

## Steps to reproduce this issue

1. Deploy a MySQL container
1.1 `docker run --name mysql -p 3306:3306 -e MYSQL_ROOT_PASSWORD=my-secret-pw -d mysql:latest`

2. Try to login with HeidiSQL
Network type: MySQL (TCP\IP) 
Library: libmysql.dll, 
Hostname: 127.0.0.1, 
User: root, 
Password: my-secret-pw, 
Port 3306

3. HeidiSQL is unable to login (Error Access Denied for user 'root@172.17.0.1' (using password: YES))

4. Open MySQL Workbench and login with same settings as Heidi, login succeed

5. Go back to HeidiSQL and attempt login, login now succeed

6. Close MySQL Workbench, disconnect HeidiSQL and re-attempt login, login succeed
 

## Current behavior
I am able to login using HeidiSQL only after accessing first the DB from MySQL Workbench, any further attempt with HeidiSQL after the successful MySQL Workbench login succeeds.

To reproduce the issue, restart the container `docker restart mysql` and re-attempt to login with HeidiSQL, login fails.

Accessing the DB from within the container doesn't seem to "unlock" it for Heidi too:

```
C:\Users\maxir> docker exec -it mysql /bin/bash
root@f35a92c0aeef:/# mysql -uroot -pmy-secret-pw
mysql: [Warning] Using a password on the command line interface can be insecure.
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 14
Server version: 8.0.19 MySQL Community Server - GPL

Copyright (c) 2000, 2020, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql>
```

**Note:** the same behavior happens with dbForge, Heidi can connect to the DB only after a successful dbForge login.
1. Restart the container to reproduce the issue
1. Attempt login with HeidiSQL, fails
1. Attempt login with dbForge, success
1. Attempt login with HeidiSQL, success
1. Close dbForge, any subsequent login with Heidi succeed.

## Expected behavior
I'd expect to be able to access the DB directly with HeidiSQL.

## Possible solution
I'm browsing the internet for possible similar issues, I'm am first trying to pinpoint if the issue is with HeidiSQL or any other client.

I've tested the commercial dbForge client and it successfully logins to the DB "directly" without having to access the DB with Workbench first.

## Environment

Docker version 19.03.8, build afacb8b

HeidiSQL 11.0.0.5919 (64bit)

mysql  Ver 8.0.19 for Linux on x86_64 (MySQL Community Server - GPL)

Workbench 8.0

* **HeidiSQL version:** 
I've re-attempted the debugging **also** with the nightly build 11.0.0.5943 without any change in behaviour.

* **Database system and version:** 
mysql  Ver 8.0.19 for Linux on x86_64 (MySQL Community Server - GPL)

* **Operating system:** 
Windows, however Docker is running Linux containers.

**Comment by ansgarbecker** [comment]
As you get an "access denied" error with HeidiSQL, you have touched the server and network itself seems to be running, without some firewall issue.
First thing to try out is to use a different library, libmariadb.dll or the newer libmysql. Please do that and report back if there's a different error message at least.

**Comment by Maxiride** [comment]
The above mentioned report was using the library `libmysql.dll`.

Test with `libmysql-6.1.dll` crashes the application, bug report sent using the built-in send report, for convenience I'm attaching it here too (I've redacted personal informations): [bugreport.txt](https://github.com/HeidiSQL/HeidiSQL/files/4461011/bugreport.txt)

Test with `libmaria.dll` succeeds on first attempt, after this successful login I am able to login using `libmysql.dll`, instead `libmysql-6.1.dll` still crashes the application.

Every test has been made **after** restarting the container.

## Observed behaviour
The library `libmysql.dll` fails to login on a new or restarted mysql docker container.

 `libmysql-6.1.dll` crashes the application everytime regardless of container state (new or restarted) either after and before attempting a login using the other libraries.

`libmaria.dll` succeed everytime regardless of container state (new or restarted) or previous logins attempts with the other libraries.

**Comment by ansgarbecker** [comment]
* libmysql.dll is old (from MySQL 5.6.6), and mostly there to fix compatibility issues on Windows XP, which is anyway no longer supported by the newer installer, so I should probably remove it.
* the crash in libmysql 6.1 should be silenced in the next build, but that won't make it work of course, so this is also not an option
* libmariadb is working, and it's the default library for a new connection. So I suppose we can close this, as I cannot do anything in the older libraries.

**Comment by Maxiride** [comment]
I don't know how the different libraries work but if you tell me that using the mariadb library with mysql is functionally the same I'm fine with it.

From the egoistic user perspective I simply want to use the product regardless of the underlying mechanics =)

Nonetheless from the development perspective if there is anything more I can add to help debug the issue with libmysql 6.1 I'll gladly help with more info if needed.

**Comment by ansgarbecker** [comment]
Thanks for your offer. To be honest, one reason for why I introduced that library selector was that I was not able to solve some connectivity issues with one library for all MySQL/MariaDB server versions, especially SSL related issues. Here we seem to have another one, and I'm glad to have that library option.

I just thought I could update libmysql 6.1, but that could easily break issue #722 again, so I'll leave that untouched.

**Comment by evs-xsarus** [comment]
@ansgarbecker I still have this problem with HeidiSQL 11.2.0.6213 and MySQL 8.0.21 on docker. I can connect only after first establishing a connection with another client (I'm using Beekeeper-Studio for that).

I see no log file with errors, but HeidiSQL does show a popup with an unknown error

PS I'm running Wine 5.0.3 on Ubuntu.

**Comment by ansgarbecker** [comment]
HeidiSQL now hints the user about using the default library if he does not yet:

![grafik](https://user-images.githubusercontent.com/7986591/154803779-deb08969-f7ef-40ca-ab16-1213fd889a3b.png)

That hint was yet only there if the message contained some known SSL message parts, and now for every failing connection. Also, this hint is included in the old-styled WinXP error dialog as well, and likely on Wine (which is seen as XP as far as I know).

**Comment by evs-xsarus** [comment]
I'm already using libmariadb.dll.

To summarize:

libmariadb.dll gives 'Unknown MySQL error'
libmysql-6.1.dll gives 'SSL connection error: unknown error number'
libmysql.dll gives 'Access denied for user xx@1.2.3.4'

**Comment by ansgarbecker** [comment]
@evs-xsarus please also try the latest nightly build (Help > Check for updates). The newer builds activate the `CLIENT_PLUGIN_AUTH_LENENC_CLIENT_DATA` option - that potentially fixes your issue.

**Comment by evs-xsarus** [comment]
@ansgarbecker just tried r6418 without success. The same messages as I mentioned before are thrown.

**Comment by ansgarbecker** [comment]
You could try the [latest installer](https://www.heidisql.com/installers/HeidiSQL_11.3.0.6414_Setup.exe), that updates libmariadb.dll and plugins. Not sure if that makes a difference.

**Comment by evs-xsarus** [comment]
@ansgarbecker it seems like it doesn't make a difference. Is there some debugging I can activate so I can give logs?

**Comment by ansgarbecker** [comment]
@evs-xsarus yes you can activate Debug logging in Tools > Preferences > Logging. But I suspect that won't show relevant stuff.

**Comment by evs-xsarus** [comment]
Well, not much more, but maybe the library versions might help:

```
/* 2022-02-22 11:47:37 [mysql80 (dev - lokaal)] */ /* Loading library file C:\Program Files\HeidiSQL\libmysql-6.1.dll ... */
/* 2022-02-22 11:47:37 [mysql80 (dev - lokaal)] */ /* C:\Program Files\HeidiSQL\libmysql-6.1.dll v6.1.11 loaded. */
/* 2022-02-22 11:47:37 [mysql80 (dev - lokaal)] */ /* Connecting to 127.0.0.1 via MariaDB or MySQL (TCP/IP), username root, using password: Yes ... */
/* 2022-02-22 11:47:37 [mysql80 (dev - lokaal)] */ /* SSL connection error: unknown error number

Please select a different library in your session settings. (Current: "libmysql-6.1.dll") */
/* 2022-02-22 11:47:44 [mysql80 (dev - lokaal)] */ /* Loading library file C:\Program Files\HeidiSQL\libmariadb.dll ... */
/* 2022-02-22 11:47:44 [mysql80 (dev - lokaal)] */ /* C:\Program Files\HeidiSQL\libmariadb.dll v3.2.5 loaded. */
/* 2022-02-22 11:47:44 [mysql80 (dev - lokaal)] */ /* Connecting to 127.0.0.1 via MariaDB or MySQL (TCP/IP), username root, using password: Yes ... */
/* 2022-02-22 11:47:44 [mysql80 (dev - lokaal)] */ /* Unknown error */
/* 2022-02-22 11:47:55 [mysql80 (dev - lokaal)] */ /* Loading library file C:\Program Files\HeidiSQL\libmysql.dll ... */
/* 2022-02-22 11:47:55 [mysql80 (dev - lokaal)] */ /* C:\Program Files\HeidiSQL\libmysql.dll v5.6.6-m9 loaded. */
/* 2022-02-22 11:47:55 [mysql80 (dev - lokaal)] */ /* Connecting to 127.0.0.1 via MariaDB or MySQL (TCP/IP), username root, using password: Yes ... */
/* 2022-02-22 11:47:55 [mysql80 (dev - lokaal)] */ /* Access denied for user 'root'@'172.18.0.1' (using password: YES) */
```

**Comment by ansgarbecker** [comment]
That states you have the right libraries. Please once again update to the latest build to be sure.

**Comment by evs-xsarus** [comment]
Heidi 6418 with the libraries from the 6414 installer also fail.

**Comment by ansgarbecker** [comment]
Shit. Probably Wine is doing something weird with the library. Can you test from a Win7/10/11 box?

**Comment by evs-xsarus** [comment]
Sadly no. In the opening comment, it seems that @maxiride uses a Windows machine. Maybe he can test?

**Comment by maxiride** [comment]
> Sadly no. In the opening comment, it seems that @maxiride uses a Windows machine. Maybe he can test?

Yes I work on a Windows machine, I've skimmed through the conversation to get up to speed on the news on the subject. 

I've been using libmaria with no troubles since my last comment, take in account that I use Heidi for quick and simple operations, so I am not definitively squeezing everything out of it since I use DataGrip for more involved operations.

I've updated Heidi to 11.3.0.6418 and I am able to login to a MySQL 5.7 instance using all the libraries on a first attempt. So I can't reproduce the issue you are now having @evs-xsarus.

If you want me to perform some specific tests, I am available.

**Comment by evs-xsarus** [comment]
Do you have a docker with MySQL 8.0 available? As that's the setup which isn't working, only after connecting to it with another client first (like Beekeeper).

**Comment by maxiride** [comment]
I've tested the three libraris, before each test I nuked and launched a new docker container:
`docker run --name mysql8-heidi-test -p 3308:3306 -e MYSQL_ROOT_PASSWORD=pass --rm mysql:8`

- libmaria OK
- libmysql-6.1 **KO**, the new message hint is showed correctly (although localized for me, suggesting to use libmaria.dll)
![image](https://user-images.githubusercontent.com/3630857/155507133-31cc5085-ad21-4f88-a45b-b6c7c302a94a.png)
- libmysql **KO**
![image](https://user-images.githubusercontent.com/3630857/155507312-8dea7abc-2f9d-43c2-b8c9-31e24b79a315.png)
Docker logs shows the error: 
```
[...] /truncated
2022-02-24T10:27:11.892947Z 0 [System] [MY-010931] [Server] /usr/sbin/mysqld: ready for connections. Version: '8.0.28'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  MySQL Community Server - GPL.
mbind: Operation not permitted
```


---

I've also tried the previous workaround of logging in with another way (DataGrip) and then attempt a login with heidi using libmysql-6.1 and libmysql but both still failed with the above errors.

**Comment by evs-xsarus** [comment]
Note, the connection worked after selecting 'Use SLL' on the SSL-tab for a connection.
